src/Compilers/CSharp/Portable/Symbols/Symbol.cs

```diff
@@ -346,6 +346,11 @@ internal static ImmutableArray<SyntaxReference> GetDeclaringSyntaxReferenceHelpe
         /// </summary>
         public abstract bool IsStatic { get; }
 
+        /// <summary>
+        /// Returns true if this symbol requires an instance reference as the implicit reciever. This is false if the symbol is static, or a <see cref="LocalFunctionSymbol"/>
+        /// </summary>
+        public virtual bool RequiresInstanceReceiver => !IsStatic;
```

Comment:
It feels like this is probably the wrong implementation approach. I do think that the public `ISymbol.IsStatic` should reflect what this PR achieves, but there's a lot of internal code that depends on the implementation of `IsStatic`. A less risky approach would be to explicitly implement the public version of `IsStatic` separately from the internal `Symbol.IsStatic`, and doc what the difference is.

Reply:
In the discussion you quoted, I think @agocke argued that:

> I can follow along this logic, but the PR looks like ISymbol.IsStatic will return the old behavior (always true). How are we rationalizing the difference between the public and private API? Why should our private API return something different from our public API?
> If the argument is that this is a breaking change because it’s the right thing to do, shouldn’t we “put our money where our mouth is” and change the behavior in the internal compiler implementation as well?

It doesn't look like that was ever resolved.

I'm happy to go either way once a decision is madr

Reply:
@agocke is oof for the next couple of weeks, so @gafter would you care to weigh in? I think the real question here is whether we should be pragmatic and attempt to change the compiler implementation as little as possible, or whether we should attempt to adapt our use cases and potentially miss some?

Reply:
I must admit I'd be a little worried about:

> A less risky approach would be to explicitly implement the public version of IsStatic separately from the internal Symbol.IsStatic, and doc what the difference is.

It does reduce risk _now_, but I'd then be worried if our public API was getting the same attention of the internal code. Or also if the compiler at one point would accidentally call the ISymbol implementation and get the "wrong" behavior.

Reply:
@333fred @agocke @gafter Which approach do we want to take here? I think @YairHalberstadt is blocked on it at this point and we can address the lack of tests or other concerns once we sort that out.

Reply:
I haven't reviewed the PR in full yet, but I like the approach @YairHalberstadt is taking here.

My primary concern with the whole thing is that we have a very loose contract on what "IsStatic" actually means, but then we used IsStatic in a lot of the code to make very specific language determinations.

From my perspective, the right approach is to pull out a new API, like @YairHalberstadt has done with RequiresInstanceReceiver, give that a strictly defined meaning, and then use that to make the appropriate language decisions.

Given this path I would actually argue that in a perfect world we would remove `ISymbol.IsStatic` since I can not think of a description in a doc comment that would be useful to a user, but clearly that ship has already sailed.